AC3-2-Final-Retake-Fireblog/AC3-2-Final-Retake-Fireblog/FeedViewController.swift

```diff
@@ -0,0 +1,220 @@
+//
+//  FeedViewController.swift
+//
+//
+//  Created by Erica Y Stevens on 4/13/17.
+//
+//
+
+import UIKit
+import Firebase
+
+class FeedViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    
+    // MARK: Stored Properties
+    
+    fileprivate var _authHandle: FIRAuthStateDidChangeListenerHandle!
+    fileprivate var _refHandle: FIRDatabaseHandle!
+    fileprivate var _storageHandle: FIRStorageHandle!
+    var databaseReference: FIRDatabaseReference!
+    var storageReference: FIRStorageReference!
+    var user: FIRUser?
+    var postInfoFromDatabase = [String:[String:Any]]()
+    var posts = [Post]()
+    
+    // MARK: View Life Cycle
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.view.backgroundColor = .white
+        
+        checkForLoggedInUser()
+        setupViewHiearchy()
+        configureConstraints()
+        configureTableView()
+        getPostsFromDatabase { (postInfoDict) in
+            DispatchQueue.main.async {
+                self.buildPostObjectsFromDatabaseInfo(postInfoDict, completion: { (post) in
+                    self.posts.append(post)
+                    self.tableView.reloadData()
+                })
+            }
+        }
+    }
+    
+    // MARK: View Hierarchy and Constraints
+    
+    func setupViewHiearchy() {
+        self.view.addSubview(tableView)
+    }
+    
+    func configureConstraints() {
+        let _ = [
+            tableView.widthAnchor.constraint(equalTo: self.view.widthAnchor),
+            tableView.heightAnchor.constraint(equalTo: self.view.heightAnchor),
+            tableView.centerXAnchor.constraint(equalTo: self.view.centerXAnchor),
+            tableView.centerYAnchor.constraint(equalTo: self.view.centerYAnchor)
+            ].map{ $0.isActive = true }
+    }
+    
+    // MARK: TableView Configuration
+    
+    func configureTableView() {
+        self.tableView.delegate = self
+        self.tableView.dataSource = self
+        self.tableView.register(FeedTableViewCell.self, forCellReuseIdentifier: "FeedCell")
+        self.tableView.estimatedRowHeight = 100
+        self.tableView.rowHeight = UITableViewAutomaticDimension
+        self.tableView.separatorStyle = .none
+    }
+    
+    // MARK: Firebase
+    
+    func checkForLoggedInUser() {
+        _authHandle = FIRAuth.auth()?.addStateDidChangeListener({ (auth: FIRAuth?, user: FIRUser?) in
```

Comment:
I'm not sure whether this listener is necessary since you have it in the AppDelegate as well. I'm also not sure if there's any harm in this being redundant but my concern is that having too many "listeners" open at once in your code calls for more cleanup. No points taken off; just be sure this is a conscious decision.